src/nupic/algorithms/TemporalMemory.cpp

```diff
@@ -0,0 +1,1093 @@
+  Permanence permanenceIncrement,
+  Permanence permanenceDecrement,
+  Int seed)
+{
+  initialize(
+    columnDimensions,
+    cellsPerColumn,
+    activationThreshold,
+    initialPermanence,
+    connectedPermanence,
+    minThreshold,
+    maxNewSynapseCount,
+    permanenceIncrement,
+    permanenceDecrement,
+    seed);
+}
+
+TemporalMemory::~TemporalMemory()
+{
+}
+
+void TemporalMemory::initialize(
+  vector<UInt> columnDimensions,
+  Int cellsPerColumn,
+  Int activationThreshold,
+  Permanence initialPermanence,
+  Permanence connectedPermanence,
+  Int minThreshold,
+  Int maxNewSynapseCount,
+  Permanence permanenceIncrement,
+  Permanence permanenceDecrement,
+  Int seed)
+{
+  // Error checking
+  if (columnDimensions.size() <= 0)
+    NTA_THROW << "Number of column dimensions must be greater than 0";
+
+  if (cellsPerColumn <= 0)
+    NTA_THROW << "Number of cells per column must be greater than 0";
+
+  // TODO: Validate all parameters (and add validation tests)
+
+  // Save member variables
+
+  numColumns_ = 1;
+  columnDimensions_.clear();
+  for (auto & columnDimension : columnDimensions)
+  {
+    numColumns_ *= columnDimension;
+    columnDimensions_.push_back(columnDimension);
+  }
+
+  cellsPerColumn_ = cellsPerColumn;
+  activationThreshold_ = activationThreshold;
+  initialPermanence_ = initialPermanence;
+  connectedPermanence_ = connectedPermanence;
+  minThreshold_ = minThreshold;
+  maxNewSynapseCount_ = maxNewSynapseCount;
+  permanenceIncrement_ = permanenceIncrement;
+  permanenceDecrement_ = permanenceDecrement;
+
+  // Initialize member variables
+  connections = Connections(numberOfCells());
+  seed_((UInt64)(seed < 0 ? rand() : seed));
+
+  activeCells.clear();
+  predictiveCells.clear();
+  activeSegments.clear();
+  winnerCells.clear();
+}
+
+void TemporalMemory::compute(UInt activeColumnsSize, UInt activeColumns[], bool learn)
+{
+  set<Cell> _activeCells(activeCells);
+  set<Cell> _winnerCells(winnerCells);
+  vector<Segment> _activeSegments(activeSegments);
+  set<Cell> _predictiveCells(predictiveCells.begin(), predictiveCells.end());
+  set<UInt> _predictedColumns;
+
+  tie(_activeCells, _winnerCells, _activeSegments,
+    _predictiveCells, _predictedColumns)
+    = computeFn(
+      activeColumnsSize,
+      activeColumns,
+      _predictiveCells,
+      _activeSegments,
+      _activeCells,
+      _winnerCells,
+      connections,
+      learn);
+
+  activeCells = _activeCells;
+  winnerCells = _winnerCells;
+  activeSegments = _activeSegments;
+  predictiveCells.clear();
+  for (Cell c : _predictiveCells)
+    predictiveCells.push_back(c);
+}
+
+tuple<set<Cell>, set<Cell>, vector<Segment>, set<Cell>, set<UInt>>
+TemporalMemory::computeFn(
+  UInt activeColumnsSize,
+  UInt activeColumns[],
+  set<Cell>& prevPredictiveCells,
+  vector<Segment>& prevActiveSegments,
+  set<Cell>& prevActiveCells,
+  set<Cell>& prevWinnerCells,
+  Connections& _connections,
+  bool learn)
+{
+  set<UInt> _activeColumns;
```

Comment:
It was a while since I looked at this, but recall it relates to in-memory changes to those given sets and helped out with the copied data. This, and other data manipulations, are rife for optimization.